runtime/cert/cert.go

```diff
@@ -0,0 +1,86 @@
+/*
+Copyright 2021 The Flux authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cert
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"errors"
+	"net/http"
+
+	corev1 "k8s.io/api/core/v1"
+)
+
+const (
+	ClientCert = "certFile"
+	ClientKey  = "keyFile"
+	CACert     = "caFile"
+)
+
+// TlsConfigFromSecret returns a tls config created from the content of the secret.
+// If the secret does not contain either a client cert and key or ca cert an
+// error will be returned. If the secret contains only the cert or key an error
+// will be returned.
+func TlsConfigFromSecret(certSecret *corev1.Secret) (*tls.Config, error) {
+	validSecret := false
+	tlsConfig := &tls.Config{}
+
+	clientCert, clientCertOk := certSecret.Data[ClientCert]
+	clientKey, clientKeyOk := certSecret.Data[ClientKey]
+	if clientCertOk && !clientKeyOk {
```

Comment:
I would usually put mutually exclusive conditions using the same variables like these in a `switch`, because it looks a bit like a logic table:

```
switch {
  case clientCertOk && !clientKeyOk:
    // ...
  case !clientCertOk && clientKeyOk:
    // ...
  case clientCertOk && clientKeyOk:
    // ...
}
```

Modulo the error message, all you care about is that they are both present or neither, so you could also do this:

```
if clientKeyOk != clientCertOk {
  return nil, fmt.Errorf("found one of %s or %s, and expected both or neither", ClientCert, ClientKey)
}
if clientKeyOk && clientCertOk {
  // ...
}
```

If you don't think either of these an improvement, your formulation is totally fine -- just fine-tuning :-)

Reply:
You are right and I think the second option is smart.